File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/framework_demo/demo1/route_context/service/PipelineListGenerator.kt
package fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.service

import fuookami.ospf.kotlin.utils.functional.*
import fuookami.ospf.kotlin.core.frontend.model.mechanism.*
import fuookami.ospf.kotlin.framework.model.*
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.Aggregation
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.service.limits.NodeAssignmentConstraint
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.service.limits.ServiceAssignmentConstraint
import fuookami.ospf.kotlin.example.framework_demo.demo1.route_context.service.limits.ServiceCostObjective

class PipelineListGenerator(
    private val aggregation: Aggregation
) {
    operator fun invoke(): Ret<PipelineList<LinearMetaModel>> {
        val list = ArrayList<Pipeline<LinearMetaModel>>()

        list.add(NodeAssignmentConstraint(aggregation.graph.nodes, aggregation.assignment))
        list.add(ServiceAssignmentConstraint(aggregation.services, aggregation.assignment))

        list.add(ServiceCostObjective(aggregation.services, aggregation.assignment))

        return Ok(list)
    }
}

File: ospf-rust-example/src/framework/demo1/route_context/service/pipeline_list_generator.rs
//! 管道列表生成器 / Pipeline list generator

use super::super::aggregation::Aggregation;
use super::limits::{
    apply_node_assignment_constraints, apply_service_assignment_constraints,
    apply_service_cost_objective,
};
use ospf_rust_core::model::MetaModel;
use std::error::Error;

/// 路线管道列表生成器 / Route pipeline list generator
/// 对齐 Kotlin route_context/service/PipelineListGenerator
pub fn generate_pipelines(
    aggregation: &Aggregation,
    model: &mut MetaModel<f64>,
) -> Result<(), Box<dyn Error>> {
    apply_node_assignment_constraints(
        model,
        &aggregation.assignment,
        aggregation.graph.nodes.len(),
    )?;
    apply_service_assignment_constraints(
        model,
        &aggregation.assignment,
        aggregation.services.len(),
    )?;
    apply_service_cost_objective(model, &aggregation.assignment, &aggregation.services)?;
    Ok(())
}
